### PR TITLE
scheduler: delay workflow consumption

### PR DESCRIPTION
**Pull request by audrium**
closes https://github.com/reanahub/reana-server/issues/382

### CODE REVIEW

File: reana_server/config.py [hunk 1]
```diff
@@ -192,3 +192,8 @@ def _(x):
     os.getenv("REANA_SCHEDULER_SECONDS_RETRY_DELAY", "60")
 )
 """How many seconds for a workflow to be delayed once rescheduled."""
+
+REANA_SCHEDULER_SECONDS_DELAY = (
+    5 if REANA_WORKFLOW_SCHEDULING_POLICY == "balanced" else 0
+)
+"""How many seconds to wait between consuming workflows."""
```
Comment (tiborsimko) [inline reana_server/config.py:199]: I'm afraid 5 seconds may not be enough, we might need to change the paradigm of consumer looking at all possible messages regardless of arrival time when picking the next candidate to run... We can chat IRL, but here are some test data for this PR:

```console
$ kubectl exec -i -t deployment/reana-db -- psql -U reana -c 'SELECT name,run_number,owner_id,complexity,created,run_started_at,run_finished_at FROM __reana.workflow ORDER BY run_started_at'            

             name             | run_number |               owner_id               |    complexity     |          created           |       run_started_at       |      run_finished_at       
------------------------------+------------+--------------------------------------+-------------------+----------------------------+----------------------------+----------------------------
 helloworld-cwl-kubernetes    |          1 | 00000000-0000-0000-0000-000000000000 | {{1,4294967296}}  | 2021-06-21 14:53:43.542443 | 2021-06-21 14:54:02.523264 | 2021-06-21 14:54:10.481298
 slow                         |          1 | 00000000-0000-0000-0000-000000000000 | {{1,4294967296}}  | 2021-06-21 14:53:58.40507  | 2021-06-21 14:55:20.003306 | 2021-06-21 14:56:32.637216
 helloworld-yadage-kubernetes |          1 | 00000000-0000-0000-0000-000000000000 | {{1,4294967296}}  | 2021-06-21 14:54:17.071172 | 2021-06-21 14:57:35.416508 | 2021-06-21 14:57:58.044152
 fast                         |          1 | 00000000-0000-0000-0000-000000000000 | {{1,4294967296}}  | 2021-06-21 14:55:09.012428 | 2021-06-21 14:58:45.500247 | 2021-06-21 14:58:51.94583
 big                          |          1 | 00000000-0000-0000-0000-000000000000 | {{1,21474836480}} | 2021-06-21 14:54:24.309889 | 2021-06-21 14:59:55.648109 | 2021-06-21 15:00:02.102044
 helloworld-serial-kubernetes |          1 | 00000000-0000-0000-0000-000000000000 | {{1,4294967296}}  | 2021-06-21 14:58:01.045252 | 2021-06-21 15:01:05.781951 | 2021-06-21 15:01:12.244222
(6 rows)
```

```console
$ kubectl logs deployment/reana-server -c scheduler | grep priority                                    
2021-06-21 14:53:50,942 | root | MainThread | INFO | Received workflow: {'user': '00000000-0000-0000-0000-000000000000', 'workflow_id_or_name': 'helloworld-cwl-kubernetes.1', 'parameters': {'input_parameters': {}, 'operational_options': {}}, 'priority': 95, 'min_job_memory': 4294967296.0, 'retry_count': 0}
2021-06-21 14:53:50,978 | root | MainThread | INFO | Starting queued workflow: {'user': '00000000-0000-0000-0000-000000000000', 'workflow_id_or_name': 'helloworld-cwl-kubernetes.1', 'parameters': {'input_parameters': {}, 'operational_options': {}}, 'priority': 95}
2021-06-21 14:54:03,620 | root | MainThread | INFO | Received workflow: {'user': '00000000-0000-0000-0000-000000000000', 'workflow_id_or_name': 'slow.1', 'parameters': {'input_parameters': {}, 'operational_options': {}}, 'priority': 10, 'min_job_memory': 4294967296.0, 'retry_count': 0}
2021-06-21 14:55:08,712 | root | MainThread | INFO | Received workflow: {'user': '00000000-0000-0000-0000-000000000000', 'workflow_id_or_name': 'slow.1', 'parameters': {'input_parameters': {}, 'operational_options': {}}, 'priority': 10, 'min_job_memory': 4294967296.0, 'retry_count': 1}
2021-06-21 14:55:08,734 | root | MainThread | INFO | Starting queued workflow: {'user': '00000000-0000-0000-0000-000000000000', 'workflow_id_or_name': 'slow.1', 'parameters': {'input_parameters': {}, 'operational_options': {}}, 'priority': 10}
2021-06-21 14:55:13,802 | root | MainThread | INFO | Received workflow: {'user': '00000000-0000-0000-0000-000000000000', 'workflow_id_or_name': 'helloworld-yadage-kubernetes.1', 'parameters': {'input_parameters': {}, 'operational_options': {'accept_metadir': True}}, 'priority': 95, 'min_job_memory': 4294967296.0, 'retry_count': 0}
2021-06-21 14:56:18,892 | root | MainThread | INFO | Received workflow: {'user': '00000000-0000-0000-0000-000000000000', 'workflow_id_or_name': 'big.1', 'parameters': {'input_parameters': {}, 'operational_options': {}}, 'priority': 74, 'min_job_memory': 21474836480.0, 'retry_count': 0}
2021-06-21 14:57:23,980 | root | MainThread | INFO | Received workflow: {'user': '00000000-0000-0000-0000-000000000000', 'workflow_id_or_name': 'helloworld-yadage-kubernetes.1', 'parameters': {'input_parameters': {}, 'operational_options': {'accept_metadir': True}}, 'priority': 95, 'min_job_memory': 4294967296.0, 'retry_count': 1}
2021-06-21 14:57:24,004 | root | MainThread | INFO | Starting queued workflow: {'user': '00000000-0000-0000-0000-000000000000', 'workflow_id_or_name': 'helloworld-yadage-kubernetes.1', 'parameters': {'input_parameters': {}, 'operational_options': {'accept_metadir': True}}, 'priority': 95}
2021-06-21 14:57:29,053 | root | MainThread | INFO | Received workflow: {'user': '00000000-0000-0000-0000-000000000000', 'workflow_id_or_name': 'big.1', 'parameters': {'input_parameters': {}, 'operational_options': {}}, 'priority': 74, 'min_job_memory': 21474836480.0, 'retry_count': 1}
2021-06-21 14:58:34,141 | root | MainThread | INFO | Received workflow: {'user': '00000000-0000-0000-0000-000000000000', 'workflow_id_or_name': 'fast.1', 'parameters': {'input_parameters': {}, 'operational_options': {}}, 'priority': 10, 'min_job_memory': 4294967296.0, 'retry_count': 0}
2021-06-21 14:58:34,164 | root | MainThread | INFO | Starting queued workflow: {'user': '00000000-0000-0000-0000-000000000000', 'workflow_id_or_name': 'fast.1', 'parameters': {'input_parameters': {}, 'operational_options': {}}, 'priority': 10}
2021-06-21 14:58:39,218 | root | MainThread | INFO | Received workflow: {'user': '00000000-0000-0000-0000-000000000000', 'workflow_id_or_name': 'helloworld-serial-kubernetes.1', 'parameters': {'input_parameters': {}, 'operational_options': {}}, 'priority': 95, 'min_job_memory': 4294967296.0, 'retry_count': 0}
2021-06-21 14:59:44,304 | root | MainThread | INFO | Received workflow: {'user': '00000000-0000-0000-0000-000000000000', 'workflow_id_or_name': 'big.1', 'parameters': {'input_parameters': {}, 'operational_options': {}}, 'priority': 74, 'min_job_memory': 21474836480.0, 'retry_count': 2}
2021-06-21 14:59:44,330 | root | MainThread | INFO | Starting queued workflow: {'user': '00000000-0000-0000-0000-000000000000', 'workflow_id_or_name': 'big.1', 'parameters': {'input_parameters': {}, 'operational_options': {}}, 'priority': 74}
2021-06-21 14:59:49,383 | root | MainThread | INFO | Received workflow: {'user': '00000000-0000-0000-0000-000000000000', 'workflow_id_or_name': 'helloworld-serial-kubernetes.1', 'parameters': {'input_parameters': {}, 'operational_options': {}}, 'priority': 95, 'min_job_memory': 4294967296.0, 'retry_count': 1}
2021-06-21 15:00:54,471 | root | MainThread | INFO | Received workflow: {'user': '00000000-0000-0000-0000-000000000000', 'workflow_id_or_name': 'helloworld-serial-kubernetes.1', 'parameters': {'input_parameters': {}, 'operational_options': {}}, 'priority': 95, 'min_job_memory': 4294967296.0, 'retry_count': 2}
2021-06-21 15:00:54,494 | root | MainThread | INFO | Starting queued workflow: {'user': '00000000-0000-0000-0000-000000000000', 'workflow_id_or_name': 'helloworld-serial-kubernetes.1', 'parameters': {'input_parameters': {}, 'operational_options': {}}, 'priority': 95}
```

As you can see, "big" (prio=74) was launched before "helloworld-serial-kubernetes" (prio=95)...

File: reana_server/scheduler.py [hunk 2]
```diff
@@ -145,7 +146,7 @@ def requeue_workflow(self, **kwargs):
                 priority=kwargs.get("priority", 0),
                 min_job_memory=kwargs.get("min_job_memory", 0),
                 retry_count=kwargs.get("retry_count", 0) + 1,
-                delay=REANA_SCHEDULER_SECONDS_RETRY_DELAY * 1000,
+                delay=0,
```
Comment (mvidalgarcia) [inline reana_server/scheduler.py:149]: If we don't need delays I guess we need to partially revert this commit? https://github.com/reanahub/reana-commons/commit/65836cdb3e9c7d23c15e2715dcd3cc75f9690c3c

**Review by mvidalgarcia — COMMENTED** [review]
The new changes introduced by this PR must come before the release commit.

File: reana_server/scheduler.py [hunk 3]
```diff
@@ -129,48 +126,17 @@ def get_consumers(self, Consumer, channel):
...
+        logging.info("Received workflow: {}".format(workflow_submission))
         workflow_min_job_memory = workflow_submission.pop("min_job_memory", 0)
-        retry_count = workflow_submission.pop("retry_count", 0)
+        workflow_submission.pop("priority", 0)
+        # FIXME: remove this once `retry_count` won't be sent from publisher
+        workflow_submission.pop("retry_count", 0)
         if reana_ready(workflow_min_job_memory):
             logging.info("Starting queued workflow: {}".format(workflow_submission))
             workflow_submission["status"] = "start"
-            priority = workflow_submission.pop("priority", 0)
```
Comment (mvidalgarcia) [inline reana_server/scheduler.py:173]: Now we can't see the priority in the logs when we start a workflow?

File: reana_server/scheduler.py [hunk 4]
```diff
@@ -28,10 +28,7 @@
     current_rwc_api_client,
     current_workflow_submission_publisher,
```
Comment (mvidalgarcia) [inline reana_server/scheduler.py:29]: This import is unused as we don't requeue ourselves anymore.

**Review by mvidalgarcia — APPROVED** [review]
Leave a couple of minor comments. It works well for me now! 👍

**Review by tiborsimko — APPROVED** [review]
Works nicely :+1:

**Pull request merged**